### ISSUE TITLE
findFeaturesSegment.py - checks for overlaps too many times

### ISSUE THREAD

**Issue by lwellerastro** [DOI-USGS/ISIS3#5540]
**ISIS version(s) affected**: isis8.2.0-RC1

**Description**  
The findFeaturesSegment.py script runs findimageoverlaps and overlapstats more times than necessary which results in excessive run times and 3 to 4 times (more?) the amount of memory required. This is an enormous issue for images overlapping dozens/hundreds/thousands of other images.

I sent several thousand images/jobs to cluster running the script and asked for 50G of memory, 7 cpus and several days of walltime for each job - far more than what findfeatures has needed for Kaguya TC polar images with many thousands of overlaps. Only 200+ jobs made it to the que (killed the remaining jobs when I saw what the problem was) and only 7 ran to completion. All other jobs either hit the walltime or exceeded memory.

After carefully looking at some of the successes, it became clear the overlap checks were being run more times than necessary. One successful image had 48 images in the fromlist (already pre-screened for overlaps via a different process), and findimageoverlaps was executed 192 times. That's exactly 4 times the number of overlapping images. Which seems to indicate to me that findimageoverlaps is being run for each image in the fromlist for the four possible scenarios for these data that result in two segments for each image:

> match segment 1 to fromlist segment 1 overlaps
> match segment 1 to fromlist segment 2 overlaps
> match segment 2 to fromlist segment 1 overlaps
> match segment 2 to fromlist segment 2 overlaps

Now maybe this has to be this way logically, however, based on files found under /tmp for some failure images, the list of images being sent to findimageoverlaps contains all of the images in the fromlist, not the match segment and one fromlist image segment for which it is trying to  determine overlaps for. Said differently, it looks like the images in the fromlist are being looped over in the script to determine overlaps, but all of the images in the fromlist are being passed to findimageoverlaps, not just the match file and the image being worked on in the loop.

I don't really understand exactly what the script is doing, but it seems like the intent is send a pair of image segments to findimageoverlaps/overlapstats (e.g. match segment 1 and from image segment 1) for overlap inspection, but instead sends the entire list of fromlist segments. I'm having a difficult time understanding what the script is doing and files are being written to /tmp and being overwritten with every call to findimageoverlaps/overlapstats via generic name, so it is difficult to know what's in these files at any one time (except in the case where the file exists and the job is killed, but only for one or two instances of findimageoverlaps). 

It's impossible to get results for a large number of images this way.

**How to reproduce**  
Because of all the image segments created for any one image result in many large files, I can't copy what's on scratch to my work area (Tb of storage necessary for a single image with a lot of overlaps). One of the successful runs (and mentioned above as having 192 calls to findimageovleraps) can be found under my scratch area /LRONAC/Malapert/FindFeatures/M1096602927RE_Network/. See LOG_neb5_46975860-21.out for the captured standard output.
The command line can be found in the directory specified in the file M1096602927RE_ff.log.

There is also a three image test that might be easier for debugging also under my scratch area /LRONAC/FFSeg/ThreeImages/FindFeatures/M121069057RE_Network/. The command line for that image is in the file M121069057RE_ff.log and the standard out is captured in LOG_neb12_46992834-1.out. The LOG file shows 8 runs of findfimageoverlaps/overlapstats for a match file that overlaps two other images. 

**Possible Solution**  
It's hard to make a suggestion when I don't fully understand the logic being used in the script. I understand the number of segments to be matched is data and user parameter dependent, but it seems that can be evaluated before checking for overlaps and used to determine the number of times findimageoverlaps/overlapstats needs to be run and what lists of segments need to be passed to it. These programs (especially findimageoverlaps) are extremely inefficient and probably shouldn't have been used in the first place. 

The program isisminer can determine overlaps extremely quickly and should be considered. In fact I use it ahead of time to determine overlaps for each and every image (whole image level of course) and it uses very little memory and runs quickly for thousands of images. Like findfeatures it requires a config file and it's own set of parameters so it wouldn't necessary be trivial to use, but it could make this script better in the end.

**Additional context**  
The results for images accessing the script successful are very good. Using findfeatures for LROC NAC images is superior to any other standard isis method for creating networks because point measures are registered at the several pixel level (and often subpixel registered). Standard methods are highly dependent on spice which can offset features by many 10's to hundreds of pixels resulting in numerous false positives when passed to pointreg which creates problems down the line, particularly with jigsaw which is then tasked with identifying the problems (no its job). Getting findfeatures to work on LROC NAC data is extremely important.

**Comment by Kelvinrr** [comment]
Just to update from the sprint meeting this AM, this issue and https://github.com/DOI-USGS/ISIS3/issues/5489 will be among the first to be tackled. But #5460 is higher priority for now since that is affecting potentially a lot of people.

**Comment by lwellerastro** [comment]
One thing that would greatly help in debugging/better understanding this particular issue is having the findimageoverlaps and overlapstats input/output files go to the local working directory instead of of /tmp. Not having those available is making it nearly impossible for me to really get a sense of what's going on with the portion of the script that deals with overlaps.

I tried changing the location of the temporary files to somewhere permanent, but I don't know python well enough (at all really) to get it working properly.  If someone wants to help me with those few lines of code I'd be willing to make the change to a copy of the script in my area and have a closer look at what's going on. Here's the only bit of code with line numbers where I could find /tmp references (from envs/isis8.2.0-RC1/bin/findFeaturesSegment.py) :

```
   176      # check for overlaps
   177      is_overlapping = []
   178      for image in from_images: 
   179          with tempfile.TemporaryDirectory() as tmpdir:
   180              tmpdir = Path(tmpdir)
   181              overlapfromlist = tmpdir / "fromlist.lis"
   182              overlaptolist = tmpdir / "tolist.lis"
```

**Comment by Kelvinrr** [comment]
Seems like the crux of the problem comes from the portion of the code that is computing the cartesian product. 


Right now, the code is doing `image_sets = list(itertools.product(match_segments.values(), from_segments.values()))` which creates image pairs like `[(1,1), (1,2), (2,1), (2,2)]` which is a problem because for this application, `(1,2)` and `(2,1)` are the same thing. So we would want jobs to be paired as `[(1,1), (1,2), (2,2)]` so we don't create that extra work. 

 Some generic code that can fix this: 

```python
import itertools

x,y = (x,y) if len(x) > len(y) else (y,x)
image_cartesian = list(itertools.product(x, y))
image_sets = []
for x in image_cartesian: 
    if x[0] >= x[1]:
      image_sets.append(x)
```

This way we are not creating unnecessary image pairs. I can get this in today and have something for you to test. 

On the image path issue, I can add another parameter that allows for you to set where those overlaps are being sent.

For now, if you want to export things to a different directory, you can change that code block like this:

```python
   176      # check for overlaps
   177      is_overlapping = []
   178      for image in from_images: 
   179          with tempfile.TemporaryDirectory() as tmpdir:
   180              tmpdir = Path(tmpdir)
   181.             tmpdir = Path("/my/local/directory") # overwrite the parent directory
   182              overlapfromlist = tmpdir / "fromlist.lis"
   183              overlaptolist = tmpdir / "tolist.lis"
```

Basically overwrite the `tmpdir` variable to point somewhere else, so everything afterwards would use that new base directory for findImageOverlaps

**Comment by lwellerastro** [comment]
My original comment: `One thing that would greatly help in debugging/better understanding this particular issue is having the findimageoverlaps and overlapstats input/output files go to the local working directory instead of of /tmp. Not having those available is making it nearly impossible for me to really get a sense of what's going on with the portion of the script that deals with overlaps.`

Resolution: `On the image path issue, I can add another parameter that allows for you to set where those overlaps are being sent.`

This is a nice added option, however not only are the input/output image lists going to this directory, the image segments have been moved here as well and that is a problem when the user elects to have the files go to the default /tmp area. LROC NAC images are notoriously big (most are about 0.5G, some larger) and the /tmp area we have on local systems and the cluster is 9G max. Typical real world datasets involve hundreds to thousands of images and many hundreds of overlaps, and /tmp gets filled instantly during the image segmentation process and the script fails. The image segments should reside where they had been originally created, under the directory where the work is being done, not /tmp or the optional workdir.

As for the original post resolution, the script is now properly only calling findimageoverlaps/overlapstats up to four times based on 2 image segments per overlapping image. However, the program findimageoverlaps is not suitable for images having many overlapping images, such as near the poles (or missions having lots of data like LROC). Despite working well for test data which has been constrained to <<10 images for quick runs and easier tracking, the script is not feasible to run on real project datasets and certainly not anything near the pole.

I am currently running the script on a single image (out of several thousand covering Malapert, a candidate landing site for the Artemis mission) against 282 overlapping images on our cluster node with the greatest amount of memory, 435G (all other nodes cap out around 305G). I can see that findimageoverlaps has run on match segment1 to fromlist segment1 and took about 2 hours to do so. I don't know how much memory was used though. However, the second run of findimageoverlaps is currently running on a similar amount of image segments and now using 345G of memory (I've been watching it rise over the past hour). This is not acceptable and would never work on the other nodes. We can't process images on a single node one image at a time (and take days to do it) when there are thousands of images to gather points on.

On top of this, the output of findimageoverlaps/overlapstats on the segment 1 files was sent to findfeatures which took 30 minutes to generate a network. Leading up to all of this was the running of crop to segment the images in two parts and footprintinit to establish shape information for the segment to help determine overlaps. That preprocessing took nearly 14 hours to complete. That's 16 hours of preprocessing for a single image and it's overlaps for one segment (out of 4) for a 30 minute run of findfeatures. This will never work in a reasonable way as written for hundreds, let alone thousands of images.

I do not intend to downplay the amount of work that has gone into getting the script to a workable form. My intention is to document what resources are actually required to apply it to a single image from a current project of high importance. I think the only way to feasibly run findfeatures on LROC NAC images in volume is to address the app itself. The script is having to perform a lot of complicated tasks and asking findimageoverlaps to help with that is unfortunately not the answer.

**Comment by lwellerastro** [comment]
`I am currently running the script on a single image (out of several thousand covering Malapert, a candidate landing site for the Artemis mission) against 282 overlapping images on our cluster node with the greatest amount of memory, 435G (all other nodes cap out around 305G).`

The job maxed out memory at 435.49G out of an allocated 435.16Gn (there is some level of padding when bumping up against the limit). I was able to watch the final run findfeatures and it only required 5G of memory to run.

Despite finishing, the job did not produce a final network. Findfeatures was run two times (seg1 to seg2, then seg2 to seg2) but only the first run was successful in finding points. The second round did not find any because segment 2 of the image being worked on is all shadow. The script is not handling something properly here that I thought this issue was addressed via #5489. I think the difference is in the 5489 test case, only segment 1 for both images overlapped whereas here there are more than one set of of segments overlaps but findfeatures fails to find points some of the sets. 

Regardless, the script failed to produce a final network for a more realistic set of data and essentially used all of our cluster memory resources in trying to do so.

**Comment by astrostu** [comment]
@lwellerastro – I know my response will be unhelpful, but I'm going to make it anyway.  This was a problem I found with the FINDIMAGEOVERLAPS program about seven years ago when working with CTX.  The massive memory issues were what led me to my first foray to write an ISIS alternative script in Python, in this case using the shapely package to calculate overlaps, write them to a text file, and then my own version of AUTOSEED which used my formatted text file since I couldn't figure out how to write in hex.

It seems like, perhaps, the method itself that FINDIMAGEOVERLAPS uses, or perhaps the way it handles both RAM and disk space, might need to be rewritten for these kinds of larger projects.

**Comment by lwellerastro** [comment]
@astrostu, we've know about the inefficiencies of findimageoverlaps since its inception. It was quickly abandoned and other methods were developed including scripting to place points on images and ultimately findfeatures. 

findimageoverlaps is not the answer for getting findfeatures to work on longer images like LROC NAC (which is what this script is trying to achieve by breaking up images into smaller segments and feeding thouse to findfeatures in a very complex way). Both apps need attention, but the findfeatures suite of generating image-image point collections is FAR superior to findimageoverlaps/autoseed and our efforts should go to making the actual findfeatures app work on images longer than 32000 lines. I don't disagree that findimageoverlaps should be improved, but it is less of a priority when something much better exists (imo).
